src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -0,0 +1,3090 @@
+      /* Not a bucket, skip it */
+      errno = 0;
+      continue;
+    }
+    if (entry->d_name[0] == '.') {
+      /* Skip dotfiles */
+      errno = 0;
+      continue;
+    }
+
+    /* TODO Use stat_to_ent */
+    //RGWBucketEnt ent;
+    //ent.bucket.name = decode_name(entry->d_name);
+    //bucket_statx_save(stx, ent, mtime);
+    RGWBucketInfo info;
+    info.bucket.name = url_decode(entry->d_name);
+    info.owner.id = std::to_string(stx.stx_uid); // TODO convert to owner name
+    info.creation_time = from_statx_timestamp(stx.stx_btime);
+
+    std::unique_ptr<rgw::sal::Bucket> bucket;
+    ret = driver->get_bucket(this, info, &bucket);
+    if (ret < 0) {
+      ldpp_dout(dpp, 0) << "ERROR: could not get bucket " << info.bucket << ": "
+	<< cpp_strerror(ret) << dendl;
+      buckets.clear();
+      return -ret;
+    }
+
+    buckets.add(std::move(bucket));
+
+    errno = 0;
+  }
+  ret = errno;
+  if (ret != 0) {
+    ldpp_dout(dpp, 0) << "ERROR: could not list buckets for " << get_display_name() << ": "
+      << cpp_strerror(ret) << dendl;
+    buckets.clear();
+    return -ret;
+  }
+
+  return 0;
+}
+
+int POSIXUser::create_bucket(const DoutPrefixProvider* dpp,
+			      const rgw_bucket& b,
+			      const std::string& zonegroup_id,
+			      rgw_placement_rule& placement_rule,
+			      std::string& swift_ver_location,
+			      const RGWQuotaInfo * pquota_info,
+			      const RGWAccessControlPolicy& policy,
+			      Attrs& attrs,
+			      RGWBucketInfo& binfo,
+			      obj_version& ep_objv,
+			      bool exclusive,
+			      bool obj_lock_enabled,
+			      bool* existed,
+			      req_info& req_info,
+			      std::unique_ptr<Bucket>* bucket_out,
+			      optional_yield y)
+{
+  /* Check for existence */
+  {
+    std::unique_ptr<rgw::sal::Bucket> bucket;
+
+    int ret = driver->get_bucket(dpp, this, b, &bucket, y);
+    if (ret >= 0) {
+      *existed = true;
+      // Bucket exists.  Check owner comparison
+      if (bucket->get_info().owner.compare(this->get_id()) != 0) {
+	return -EEXIST;
+      }
+      // Don't allow changes to ACL policy
+      RGWAccessControlPolicy old_policy(driver->ctx());
+      ret = rgw_op_get_bucket_policy_from_attr(
+          dpp, driver, this, bucket->get_attrs(), &old_policy, y);
+      if (ret >= 0 && old_policy != policy) {
+        bucket_out->swap(bucket);
+        return -EEXIST;
+      }
+    } else {
+      *existed = false;
+    }
+  }
+
+  binfo.bucket = b;
+  binfo.owner = get_id();
+  binfo.zonegroup = zonegroup_id;
+  binfo.placement_rule = placement_rule;
+  binfo.swift_ver_location = swift_ver_location;
+  binfo.swift_versioning = (!swift_ver_location.empty());
+  binfo.requester_pays = false;
+  binfo.creation_time = ceph::real_clock::now();
+  if (pquota_info) {
+    binfo.quota = *pquota_info;
+  }
+
+  POSIXBucket* fb = new POSIXBucket(driver, driver->get_root_fd(), binfo, this);
+
+  int ret = fb->set_attrs(attrs);
+  if (ret < 0) {
+    delete fb;
```

Comment:
Using a unique_ptr requires jumping through hoops when assigning it to the out parameter, since it's a POSIXBucket not a Bucket.  Ugly either way, so I don't have a huge preference.

Reply:
unique_ptr move-assignment from a derived type to a base type works fine. this is overload (2) in https://en.cppreference.com/w/cpp/memory/unique_ptr/operator%3D

i wrote a trivial example in https://godbolt.org/z/rf378cvqE

Reply:
Interesting.  It certainly didn't work on the compiler I had when I wrote RADOSStore.  That's why I used this pattern.